Proposed fix for LOG4J2-2610 rollover issue:

JCgH4164838Gh792C124B5 opened the pull request:
Proposed fix for LOG4J2-2610 rollover issue:
- Fix for Windows Filesystem Tunneling issue, rapid rollover issue.
- Added FileUtils.updateFileCreationTime() method.
- Updated FileManager and RollingFileManager to allow the file creationTime attribute to be updated on rollover (when it doesn't  match initialTime).
- RollingAppenderSizeWithTimeTest.testAppender() was failing on Windows before any changes.  Rapid log rollover was causing overwrite of data in some cases (instead of writing to a new log file), resulting in lost logging messages.  Mitigation was to apply a 1ms sleep in RollingFileManager.createFileAfterRollover().

JCgH4164838Gh792C124B5 commented:
Hello Apache Log4j2 Team.

Hopefully the code in this PR is formatted appropriately, but let me know if different formatting would be preferred.  The same goes for the code logic, as the committers may want an alternative solution to wha't been proposed.

rgoers commented:
No offense, but this PR is just another example of why I hate Windows.

garydgregory commented on log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/RollingFileManager.java:
```diff
@@ -344,7 +344,12 @@ public synchronized void rollover() {
 
     protected void createFileAfterRollover() throws IOException  {
-        setOutputStream(createOutputStream());
+        try {
```
Why not set the file timestamp specifically?

JCgH4164838Gh792C124B5 replied:
The code path called by line 352, `setOutputStream(createOutputStream(initialTime))`, does attempt to set the creationTime attribute explicitly.  It uses the value of initialTime (which was set previously in `rollover()`).

The 1ms sleep on line 348 was introduced as a mitigation for `RollingAppenderSizeWithTimeTest.testAppender()` failing on Windows during its rapid rollovers (confirmed to happen before any changes in this PR were applied).  A 0ms sleep was tried but that test still failed (it's possible the 1ms sleep succeeds due to a different priority system thread associated with JVM I/O needing to execute).

Does that make sense, or did I misinterpret your question ?

carterkozak replied:
That test is flaking a lot of the linux build nodes as well as when I run the test locally on a mac. Are you confident the failure is related to windows?

JCgH4164838Gh792C124B5 replied:
I'm _not particularly confident_ the `RollingAppenderSizeWithTimeTest.testAppender()` failure is _specific to Windows_ ... I just _assumed_ that if this test was failing with Linux someone would have reported it sooner.  :wink:

If you're saying that test is "flaking a lot" on Linux and/or Mac builds too, then that probably points to the issue being more of a general JVM I/O behaviour (i.e. could impact any OS).  Maybe the 1ms sleep mitigation will be sufficient across all OS platforms ?

carterkozak replied:
If I run the test in a loop locally, the failures are resolved by adding the 1ms sleep, however I don't think we can commit that change.
Using a relatively standard logging configuration on my development machine (macbook), I can log about ~1,000,000 events to disk every second. That millisecond represents 1,000 events that aren't logged! Do we have any options that don't involve waiting?

JCgH4164838Gh792C124B5 replied:
The 1ms sleep only occurs on log rollover ... so unless a logging configuration involves _massive numbers of rollovers_ in _very short time periods_ it shouldn't be too expensive for that minimal wait ?

A good proportion of log rollovers are done on minute/hour/day thresholds (or rollover size thresholds sufficient to hold data for at least a minute), where a 1ms sleep should be fairly negligible.  The `testAppender()` test appears to be configured to perform rollovers at a very high rate (probably a lot higher than the vast majority of timed/size rollovers).  Given that the test shows actual lost log messages (due to overwrites) it _seems_ a short sleep is _safer_ than the alternative ...

If anyone can find a way to avoid even a 1ms sleep that would be neat, but it really looks as if the JVM needs a minimal pause to dispatch something in order for the old-file-move-new-file-create to "safely complete". :disappointed:

carterkozak replied:
Filed [LOG4J2-2613](https://issues.apache.org/jira/browse/LOG4J2-2613) to track flakes.

> A good proportion of log rollovers are done on minute/hour/day thresholds (or rollover size thresholds sufficient to hold data for at least a minute), where a 1ms sleep should be fairly negligible.

In those cases with only hourly/daily thresholds, it's not the worst thing in the world, however combined with size based policies it's common to roll 1gb logs every few seconds under load.

> Given that the test shows actual lost log messages (due to overwrites) it seems a short sleep is safer than the alternative ...

I totally agree that we should prioritize fixing this, dropped messages are bad. If our only options are to sleep or to drop logs, I'd take the sleep in a heartbeat. I want to be confident that we've explored all potential avenues before applying a sleep.

JCgH4164838Gh792C124B5 commented:
Hello @rgoers.

No offense taken. :smile:  

It was a bit surprising to encounter the "tunneling" filesystem behaviour, but at least there seems to be an explanation for it.

_In theory_ it's possible that some non-Windows filesystems could produce a similar behaviour (e.g. as part of a caching scheme), or even some other strange behaviour related to creationTime attributes.  So in that sense having an extra sanity check (when the underlying filesystem supports creationTime) might still have some value. :relaxed:

rgoers commented:
I seems like a better solution would be to just set the creation time attribute for the file when it is created. I am investigating doing that.

JCgH4164838Gh792C124B5 commented:
I tried to find a (non-JNI) mechanism to set the creationTime when a file is created in Java but didn't have any luck.  It will be cool if you find a way!  :smile:
Lacking that, the best idea I could come up with was to deal with the creationTime after-the-fact.  :disappointed:

rgoers commented:
I've already written the code Files.setAttribute() should do the trick. However, I first want to recreate the problem in my Windows VM without the fix so I can verify the fix works. I built log4j-core but it didn't fail. What test is failing for you?

JCgH4164838Gh792C124B5 commented:
Hello @rgoers (**apologies in advance for the long response**).

The **short answer** is _none of the existing test cases actually failed_ due to filesystem tunneling behaviour (was only noted by "manual observation" with the necessary preconditions).

`RollingAppenderSizeWithTimeTest.testAppender()` was the only test that was actually failing (which as per discussion above with @carterkozak _actually appears to be a separate issue_ - where the 1ms sleep in `RollingFileManager.createFileAfterRollover()` was needed to get the test to succeed).

When debugging an application that used Log4j2 with a "size-and-time" rollover strategy, I noted that when the _current log file_ had a _`creationTime` > 24hrs older than the current date/time_, the rollover behaved strangely.  When the current log file was rolled over the new "xxxx.log" would "_carry over_" the same >24hrs old `creationTime` as the original file.  Then, later, on the next rollover, it's rolled-over name ("xxxx_CCYY-MM-DD_y.log") still had the CCCYY-MM-DD represented by the `creationTime` of the original log file (2 rollovers ago).

Basically you need sufficient _"discontiguous time"_ between the creationTime of the original file and the actual current system datetime to observe the behaviour.  The amount of time required to observe it depended on the rollover period.

Due to the conditions required to reproduce it, I didn't initally consider a test case to try and demonstrate it failing.  Instead, I did a manual test set up using `RollingAppenderSizeWithTimeTest`.  I either suppressed the directory cleanup "`withCleanFoldersRule`" in the test, or used a breakpoint so I could copy a log file with a "really old" creationTime in place.

Using the new code with Files.setAttribute() that you mentioned, I suppose you could create an empty file in the test before the appender tries to access it.  Then you could set its creationTime to >24hrs in the past, let the appender roll it over/create the new .log file and then _confirm by reading its creationTime attribute that it is using "today's date" as expected_ ?

**Sorry again for the long explanation**, but it really is/was a pain to reproduce in a "nice" way.  :sweat_smile:.

rgoers commented:
I modified RollingAppenderTimeAndSizeTest and was able to detect the problem on Windows. I was then able to verify that my fix worked. If you agree please close this PR.

JCgH4164838Gh792C124B5 commented:
Hello @rgoers .

I updated to the latest release-2.x and re-ran the regression.  Your change and updated unit test passed for me. 👍 

However, the other issue identified in this PR (item 4 in the summary text) and mentioned in the associated Jira still causes RollingAppenderSizeWithTimeTest to fail:

```
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.525 s <<< FAILURE! - in org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest
[ERROR] testAppender(org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest)  Time elapsed: 6.525 s  <<< FAILURE!
java.lang.AssertionError: Log messages lost : 217
	at org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest.testAppender(RollingAppenderSizeWithTimeTest.java:95)
```

So far the 1ms sleep in RollingFileManager.createFileAfterRollover() (also in this PR) is the only mitigation found that works.  This failure is being tracked by LOG4J2-2613, so is it OK to leave this PR open until a better solution is determined ?

rgoers commented:
I ran RollingAppenderSizeWithTimeTest in my Windows VM the other day and it failed. I am trying to diagnose the problem today and have run it many times and cannot get it to fail.

carterkozak commented:
@rgoers I've been able to get the test to fail locally with this: #277
That said there could be something odd going on with the test itself that makes repeated runs fail in a way that doesn't represent the original flake.

rgoers commented:
@carterkozak Even with the Suite I still cannot get it to fail.

Can you apply the attached diff and see if it fails for you? If it does it should generate some logging that may help diagnose the problem.
[diff.txt](https://github.com/apache/logging-log4j2/files/3294575/diff.txt)

carterkozak commented:
I think we have a subtle race somewhere, when I keep `statusLogger.registerListener(this);` commented I can reproduce failures, but when I uncomment it for more debugging information I cannot repro the failure anymore.

carterkozak commented:
If I set the status logger level to trace and don't register the listener, I don't see anything helpful in the log surrounding failures:
```
DEBUG StatusLogger Using configurationFactory org.apache.logging.log4j.core.config.ConfigurationFactory$Factory@36d4b5c
TRACE StatusLogger Trying to find [log4j-rolling-size-with-time.xml] using context class loader sun.misc.Launcher$AppClassLoader@18b4aac2.
DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger Closing BufferedInputStream java.io.BufferedInputStream@53b32d7
DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger PluginManager 'Converter' found 46 plugins
DEBUG StatusLogger Starting OutputStreamManager SYSTEM_OUT.false.false-8
DEBUG StatusLogger Using configurationFactory org.apache.logging.log4j.core.config.ConfigurationFactory$Factory@36d4b5c
TRACE StatusLogger Trying to find [log4j-rolling-size-with-time.xml] using context class loader sun.misc.Launcher$AppClassLoader@18b4aac2.
DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger Closing BufferedInputStream java.io.BufferedInputStream@3abbfa04

java.lang.AssertionError: Log messages lost : 32

	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest.testAppender(RollingAppenderSizeWithTimeTest.java:116)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.apache.logging.log4j.junit.LoggerContextRule$1.evaluate(LoggerContextRule.java:123)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger PluginManager 'Converter' found 46 plugins
DEBUG StatusLogger Starting OutputStreamManager SYSTEM_OUT.false.false-9
DEBUG StatusLogger Using configurationFactory org.apache.logging.log4j.core.config.ConfigurationFactory$Factory@36d4b5c
TRACE StatusLogger Trying to find [log4j-rolling-size-with-time.xml] using context class loader sun.misc.Launcher$AppClassLoader@18b4aac2.
DEBUG StatusLogger Not in a ServletContext environment, thus not loading WebLookup plugin.
DEBUG StatusLogger Closing BufferedInputStream java.io.BufferedInputStream@73ad2d6
```

rgoers commented:
Yes, I am pretty sure this is timing related. I just don't know if it is something in the test or something in the code. 

If you had trace enabled you should have seen logs from when the files roll over. The reason for registering the listener is it will just store the log event in a queue and will only print it if the error occurs. Enabling the logging will almost certainly impact the test somehow as writing to the console is pretty slow.

carterkozak commented:
You're right, I set the status logger level too early and failed to record anything while the test was running.
I was not able to reproduce a failure with trace enabled, but the one iteration failed when I set the level to debug:

```
2019-06-16 20:35:27,626 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,626 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,627 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-27-626.log, renameEmptyFiles=false]
2019-06-16 20:35:27,627 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:27.627-0400
2019-06-16 20:35:27,628 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,628 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,628 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-27-628.log, renameEmptyFiles=false]
2019-06-16 20:35:27,629 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:27.629-0400
2019-06-16 20:35:27,629 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,629 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:27,630 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-27-629.log, renameEmptyFiles=false]
2019-06-16 20:35:27,630 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:27.630-0400
2019-06-16 20:35:28,132 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,132 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,133 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-132.log, renameEmptyFiles=false]
2019-06-16 20:35:28,133 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.133-0400
2019-06-16 20:35:28,133 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,134 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,134 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-134.log, renameEmptyFiles=false]
2019-06-16 20:35:28,134 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.134-0400
2019-06-16 20:35:28,134 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,134 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,134 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-134.log, renameEmptyFiles=false]
2019-06-16 20:35:28,135 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.135-0400
2019-06-16 20:35:28,640 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,640 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,640 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-640.log, renameEmptyFiles=false]
2019-06-16 20:35:28,641 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.640-0400
2019-06-16 20:35:28,641 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,642 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,642 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-642.log, renameEmptyFiles=false]
2019-06-16 20:35:28,642 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.642-0400
2019-06-16 20:35:28,643 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,643 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:28,643 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-28-643.log, renameEmptyFiles=false]
2019-06-16 20:35:28,644 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:28.643-0400
2019-06-16 20:35:29,146 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,147 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,147 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-147.log, renameEmptyFiles=false]
2019-06-16 20:35:29,147 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.147-0400
2019-06-16 20:35:29,148 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,148 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,149 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-148.log, renameEmptyFiles=false]
2019-06-16 20:35:29,149 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.149-0400
2019-06-16 20:35:29,150 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,150 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,150 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-150.log, renameEmptyFiles=false]
2019-06-16 20:35:29,150 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.150-0400
2019-06-16 20:35:29,653 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,653 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,653 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-653.log, renameEmptyFiles=false]
2019-06-16 20:35:29,654 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.654-0400
2019-06-16 20:35:29,655 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,655 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,655 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-655.log, renameEmptyFiles=false]
2019-06-16 20:35:29,655 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.655-0400
2019-06-16 20:35:29,656 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,656 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:29,656 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-29-656.log, renameEmptyFiles=false]
2019-06-16 20:35:29,657 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:29.656-0400
2019-06-16 20:35:30,158 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,158 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,158 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-158.log, renameEmptyFiles=false]
2019-06-16 20:35:30,159 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.159-0400
2019-06-16 20:35:30,160 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,160 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,160 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-160.log, renameEmptyFiles=false]
2019-06-16 20:35:30,161 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.160-0400
2019-06-16 20:35:30,161 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,161 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,161 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-161.log, renameEmptyFiles=false]
2019-06-16 20:35:30,162 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.162-0400
2019-06-16 20:35:30,663 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,663 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,663 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-663.log, renameEmptyFiles=false]
2019-06-16 20:35:30,664 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.664-0400
2019-06-16 20:35:30,664 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,664 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,664 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-664.log, renameEmptyFiles=false]
2019-06-16 20:35:30,665 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.665-0400
2019-06-16 20:35:30,665 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,665 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:30,665 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-30-665.log, renameEmptyFiles=false]
2019-06-16 20:35:30,666 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:30.665-0400
2019-06-16 20:35:31,171 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,172 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,172 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-172.log, renameEmptyFiles=false]
2019-06-16 20:35:31,172 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.172-0400
2019-06-16 20:35:31,173 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,173 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,173 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-173.log, renameEmptyFiles=false]
2019-06-16 20:35:31,174 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.174-0400
2019-06-16 20:35:31,174 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,174 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,175 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-174.log, renameEmptyFiles=false]
2019-06-16 20:35:31,175 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.175-0400
2019-06-16 20:35:31,677 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,677 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,677 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-677.log, renameEmptyFiles=false]
2019-06-16 20:35:31,677 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.677-0400
2019-06-16 20:35:31,678 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,678 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,678 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-678.log, renameEmptyFiles=false]
2019-06-16 20:35:31,678 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.678-0400
2019-06-16 20:35:31,679 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,679 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,679 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-679.log, renameEmptyFiles=false]
2019-06-16 20:35:31,679 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.679-0400
2019-06-16 20:35:31,680 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,680 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:31,680 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-31-680.log, renameEmptyFiles=false]
2019-06-16 20:35:31,680 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:31.680-0400
2019-06-16 20:35:32,186 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,187 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,187 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-32-187.log, renameEmptyFiles=false]
2019-06-16 20:35:32,188 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:32.187-0400
2019-06-16 20:35:32,188 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,189 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,189 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-32-189.log, renameEmptyFiles=false]
2019-06-16 20:35:32,189 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:32.189-0400
2019-06-16 20:35:32,190 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,190 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:35:32,190 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-35-32-190.log, renameEmptyFiles=false]
2019-06-16 20:35:32,190 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:35:32.190-0400
2019-06-16 20:35:32,191 main DEBUG Stopping LoggerContext[name=testAppender(org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest), org.apache.logging.log4j.core.LoggerContext@35cabb2a]...
2019-06-16 20:35:32,191 main DEBUG Shutting down RollingFileManager target/rolling-size-test/rolling.log
2019-06-16 20:35:32,192 main DEBUG Shutting down RollingFileManager target/rolling-size-test/rolling.log
2019-06-16 20:35:32,192 main DEBUG All asynchronous threads have terminated
2019-06-16 20:35:32,192 main DEBUG RollingFileManager shutdown completed with status true
2019-06-16 20:35:32,192 main DEBUG Shut down RollingFileManager target/rolling-size-test/rolling.log, all resources released: true
2019-06-16 20:35:32,192 main DEBUG Appender RollingFile stopped with status true
2019-06-16 20:35:32,192 main DEBUG Stopped XmlConfiguration[location=/Volumes/git/logging-log4j2/log4j-core/target/test-classes/log4j-rolling-size-with-time.xml] OK
2019-06-16 20:35:32,192 main DEBUG Stopped LoggerContext[name=testAppender(org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest), org.apache.logging.log4j.core.LoggerContext@35cabb2a] with status true

java.lang.AssertionError: Log messages lost : 32

	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.apache.logging.log4j.core.appender.rolling.RollingAppenderSizeWithTimeTest.testAppender(RollingAppenderSizeWithTimeTest.java:113)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.apache.logging.log4j.junit.LoggerContextRule$1.evaluate(LoggerContextRule.java:123)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

carterkozak commented:
Ah, ran several more iterations and got a failure with trace enabled: https://gist.github.com/carterkozak/c6a3bc6cfc258744bb5feb8a64129500

rgoers commented:
Thanks. That is perfect and is exactly what I expected. From the log we see 
`2019-06-16 20:39:01,573 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-39-01-573.log, renameEmptyFiles=false]
2019-06-16 20:39:01,573 main TRACE Renamed file /Volumes/git/logging-log4j2/log4j-core/target/rolling-size-test/rolling.log to /Volumes/git/logging-log4j2/log4j-core/target/rolling-size-test/rollingtest-2019-06-167T20-39-01-573.log with Files.move
2019-06-16 20:39:01,573 main DEBUG Now writing to target/rolling-size-test/rolling.log at 2019-06-16T20:39:01.573-0400
2019-06-16 20:39:01,573 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:39:01,573 main DEBUG Formatting file name. useCurrentTime=false, currentFileTime=0, prevFileTime=0
2019-06-16 20:39:01,574 main TRACE DefaultRolloverStrategy.purge() took 0.0 milliseconds
2019-06-16 20:39:01,574 main DEBUG RollingFileManager executing synchronous FileRenameAction[target/rolling-size-test/rolling.log to target/rolling-size-test/rollingtest-2019-06-167T20-39-01-573.log, renameEmptyFiles=false]`

So the file rolled over twice within the same millisecond. The second rollover overwrote the first. The easy fix for this is to add %i to the filename pattern. The other way to do it is add calls to sleep in the logging loop.

rgoers commented:
@carterkozak Please try the latest change I made on the release-2.x branch and see if it addresses the issue for you. I simply increased the size of the file required before a rollover can take place. I suspect that the real problem here is that the clock on Windows has a granularity larger than 1ms so that the rollovers have to occur at least some number of milliseconds apart to make sure the clock is updated and duplicate files don't occur.

carterkozak commented:
I was able to reproduce this failure on a mac, so I don't think it had to do with windows. I'll give it a run and check if I can still reproduce the issue. Thanks Ralph!

carterkozak commented:
@rgoers Tests all pass, I cannot reproduce the failure anymore. Thanks!

rgoers commented:
Closing this PR since there is nothing left to do.

JCgH4164838Gh792C124B5 commented:
Hello Again.

Thanks for looking into this issue.  

One last comment: It seems like the test case was updated so that the test case no longer fails.  However isn't the underlying issue of rapid-rollover still present and just masked by the test changes ?

Unless the thread calling `RollingFileManager.createFileAfterRollover()` has a slight wait/sleep the risk of the race condition may still remain.  Maybe the docs should warn that rapid-rollovers (near or less than 1ms) are not recommended as a result ?

carterkozak commented:
the Test failure was due to triggering rollover twice within a single interval of the smallest date component (millisecond) without a %i pattern present. This behavior is similar to rollover in a file with pattern “rollingtest-%d{yyyy}”, where Rollovers inside of a single year will overwrite previous rollovers. This can be solved by adding %i to the pattern to increment on each roll inside of that window, and is expected behavior. I believe the windows bug was addressed by Ralph’s commit a few days ago.

rgoers commented:
To echo Carter's comment, anytime multiple rollovers can occur within the "time window" there is a risk that the file being rolled over to will be replaced. As Carter states, simply adding %i to the pattern will prevent that as a counter will be appended to the file name. Without the counter it is up to the user to make sure the file cannot be rolled over multiple times within a time period. So using the size-based triggering policy with a file that has a timestamp but no counter will likely encounter problems whenever the file size triggers the rollover.
